[[ Bug 21659 ]] Update TreeView to ignore mouseUp when row has changed

bwmilby opened the pull request:
### Bug 21659

Add `mMouseDownRow` to record the row clicked when a mouseDown event is
registered that does not take place in the scrollBar or Separator.

Add check to the `OnClick` handler to ensure that clicks are ignored if
the mouseDown and mouseUp rows are different.

Moved the `addBaseLevelElement` handler call in `OnClick` to ensure that it is not called
when mouseDown and mouseUp rows are different.

### Other changes

Converted spaces to tabs in several places.

Added **hilitedElementFoldState** property to allow get/set of the fold state of the selected element.  This will also allow checking to see if the selected element is a leaf node.

Added `pathIsAnArray` handler to determine if the selected element is a leaf node to avoid trying to fold/unfold a leaf node when setting fold state.

### Changes from PR 6738 (included in this PR as well)

The tree view widget now has the ability to automatically expand to reveal
a row when it is selected. If `scrollHilitedElementIntoView` is not `true`
then the scroll position will be adjusted to maintain the currently visible
top row.

Added 2 module variables to control redraw and determine when fold state
changes (`mRedraw` and `mFoldChanged`).

Added `pathIsValid` handler to ensure that a user supplied path actually
exists.  Validation is now done in the `setSelectedElement` handler so the
requirement to test for an empty path in `selectPath` is removed.

Added `unfoldFullPath` handler to take care of ensuring the full selection
path is unfolded.  Code is included to maintain the top visible row if
`mScrollHilitedElementIntoView` is not `true`.

Updated `applyFoldStateKey` handler to set `mFoldChanged` when an expansion
is performed.  Updated to check for the need to expand.

### Bug 21361

**autoFoldStateReset** property added to allow the fold state to be reset
when the array data is set.

Default value is `false` to match existing behavior.  Property is saved
and restored with the other widget data.

bwmilby committed:
[[ Tree View Widget ]] Auto expand to show selected row

The tree view widget now has the ability to automatically expand to reveal
a row when it is selected. If **scrollHilitedElementIntoView** is not `true`
then the scroll position will be adjusted to maintain the currently visible
top row.

The tree view fold state can now be reset in two ways. To set the array
data in a folded state then use the **foldedArrayData** property. To
collapse the tree without changing the data, get the **foldedArrayData**.

Added 2 module variables to control redraw and determine when fold state
changes (`mRedraw` and `mFoldChanged`).

Added `pathIsValid` handler to ensure that a user supplied path actually
exists.  Validation is now done in the `setSelectedElement` handler so the
requirement to test for an empty path in `selectPath` is removed.

Added `unfoldFullPath` handler to take care of ensuring the full selection
path is unfolded.  Code is included to maintain the top visible row if
`mScrollHilitedElementIntoView` is not `true`.

Updated `applyFoldStateKey` handler to set `mFoldChanged` when an expansion
is performed.  Updated to check for the need to expand.

bwmilby committed:
[[ Bug 20659 ]] Update TreeView to ignore mouseUp when row has changed

Add `mMouseDownRow` to record the row clicked when a mouseDown event is
registered that does not take place in the scrollBar or Separator.

Add check to the `OnClick` handler to ensure that clicks are ignored if
the mouseDown and mouseUp rows are different.

Moved the `addBaseLevelElement` handler to ensure that it is not called
when mouseDown and mouseUp rows are different.

Converted spaces to tabs in several places.

bwmilby committed:
[[ Bug 21361 ]] Add option to reset fold state when setting arrayData

Add bugfix note to existing PR

montegoulding commented:
@bwmilby is this a replacement for https://github.com/livecode/livecode/pull/6738 ?

bwmilby commented:
I didn't mean it to be, but I branched it off of that one since I needed some of the changes there.  I could merge the second commit with the bugfix note into this one and make changes here (just wait until 11/1 to close the other one for Hactoberfest)

bwmilby committed:
[[ TreeView ]] Add hilitedElementFoldState property

Add a property to allow the fold state of the selected element to be
get and set.  This also allows determining if the selected element
is a leaf node.

Updated documentation for enum parameters so they are formatted properly.

Moved getSelectedElement handler so the ordering of get/set handlers is
more consistent (get handlers appear before set handlers).

bwmilby commented:
Due to how I generated the PR, I do suggest combining and working off of this one.  I’ve added the other PR note into this one.

bwmilby committed:
[[ Bug 21361 ]] Add property to allow auto reset fold state

**autoFoldStateReset** property added to allow the fold state to be reset
when the array data is set.

Default value is `false` to match existing behavior.  Property is saved
and restored with the other widget data.

bwmilby committed:
[[ TreeView Widget ]] Back out foldedArrayData property

Remove **foldedArrayData** property from widget source and feature note.

Replaced with **autoFoldStateReset** property in another commit to address
bug 21361.

PR 6770 is a better solution for manual clearing of the fold state and
more general in that it allows complete control of the fold state array.

seaniepie commented:
Dang it! I was hoping this was in the 9.0.2 release. I thought I'd finally be able to release my software. Looks like I won't get paid this year :(

montegoulding commented:
@seaniepie this PR targets the develop branch which is 9.1 as with all new features so there was never any intention of it going in a maintenance release of 9.0. That being said there's no reason at all you can't download the code, compile the widget and use it in 9.0.2 if you need it right now.

livecodeali commented:
@livecode-vulcan review ok 68bafd3

livecode-vulcan commented:
💙 review by @livecodeali ok 68bafd355bcf6960f63b007cfcf9c571d591bb8a

livecode-vulcan commented:
😞 test failure 68bafd355bcf6960f63b007cfcf9c571d591bb8a

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/393)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/385)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/385)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/385)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/385)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1907)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/135)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/134)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1938)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1928)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1927)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1959)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1900)

livecodeali commented:
@livecode-vulcan retry 68bafd3

livecode-vulcan commented:
😞 test failure 68bafd355bcf6960f63b007cfcf9c571d591bb8a

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/394)
* try-community-armv7-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/386)
* try-community-arm64-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/386)
* try-community-x86-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/386)
* try-community-x86_64-android-sdk26_ndk16r15: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/386)
* try-community-js-emscripten-sdk1.35: [pending](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1908)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/136)
* try-community-universal-ios-iphonesimulator12.1: [pending](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/135)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1939)
* try-community-x86-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1929)
* try-community-x86_64-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1928)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1960)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1901)

livecodeali commented:
@livecode-vulcan retry 68bafd3

livecode-vulcan commented:
😎 test success 68bafd355bcf6960f63b007cfcf9c571d591bb8a

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/396)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/388)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/388)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/388)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/388)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1910)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/138)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/137)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1941)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1931)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1930)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1962)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1903)
